File: test/lib/vbrain/paths_test.rb
require "test_helper"
require "vbrain/paths"

class PathsTest < Minitest::Test
  def test_data_home_uses_env_when_set
    old = ENV["VBRAIN_HOME"]
    ENV["VBRAIN_HOME"] = "/tmp/custom-vbrain"
    assert_equal "/tmp/custom-vbrain", VBrain::Paths.data_home
  ensure
    ENV["VBRAIN_HOME"] = old
  end

  # Sem VBRAIN_HOME, rodando de dentro de uma base (PROJECT_ROOT tem wiki/),
  # o código usa essa base — é o que conserta o cloud, onde o checkout é a
  # base e o sub-agente não herda VBRAIN_HOME do shell.
  def test_data_home_uses_local_base_when_env_blank_and_running_in_base
    old = ENV["VBRAIN_HOME"]
    ENV["VBRAIN_HOME"] = nil
    assert VBrain::Paths.base?(VBrain::Paths::PROJECT_ROOT),
      "pré-condição: o repo de teste é uma base (tem wiki/)"
    assert_equal VBrain::Paths::PROJECT_ROOT, VBrain::Paths.data_home
    ENV["VBRAIN_HOME"] = ""
    assert_equal VBrain::Paths::PROJECT_ROOT, VBrain::Paths.data_home
  ensure
    ENV["VBRAIN_HOME"] = old
  end

  # Sem VBRAIN_HOME e fora de uma base, cai no ~/vbrain padrão.
  def test_data_home_defaults_to_home_vbrain_when_env_blank_and_not_in_base
    old = ENV["VBRAIN_HOME"]
    ENV["VBRAIN_HOME"] = nil
    VBrain::Paths.stub :base?, false do
      assert_equal File.expand_path("~/vbrain"), VBrain::Paths.data_home
    end
  ensure
    ENV["VBRAIN_HOME"] = old
  end

  def test_derived_paths_are_under_data_home
    with_isolated_data_home do |dir|
      assert_equal File.join(dir, "raw"), VBrain::Paths.raw_dir
      assert_equal File.join(dir, "wiki"), VBrain::Paths.wiki_dir
      assert_equal File.join(dir, "db", "vbrain.sqlite3"), VBrain::Paths.db_path
      assert_equal File.join(dir, "raw", ".tmp"), VBrain::Paths.tmp_dir
    end
  end

  def test_ensure_dirs_creates_flat_structure
    with_isolated_data_home do |dir|
      VBrain::Paths.ensure_dirs!
      assert Dir.exist?(File.join(dir, "raw"))
      assert Dir.exist?(File.join(dir, "wiki"))
      assert Dir.exist?(File.join(dir, "db"))
      assert Dir.exist?(File.join(dir, "raw", ".tmp"))
      assert Dir.exist?(File.join(dir, "wiki", VBrain::Paths::REALTIME_DIR)),
        "subdir _realtime preservado"
    end
  end

  def test_ensure_dirs_does_not_create_type_folders
    with_isolated_data_home do |dir|
      VBrain::Paths.ensure_dirs!
      %w[concepts decisions gotchas notes _rules].each do |old|
        refute Dir.exist?(File.join(dir, "wiki", old)),
          "pasta de tipo #{old} não deve mais ser criada (layout plano)"
      end
    end
  end

  def test_project_root_points_to_repo
    assert VBrain::Paths::PROJECT_ROOT.end_with?("vbrain")
    assert File.exist?(File.join(VBrain::Paths::PROJECT_ROOT, "Gemfile"))
  end

  def test_kinds_include_all_supported_metadata
    %w[concept decision gotcha note rule realtime].each do |k|
      assert_includes VBrain::Paths::KINDS, k
    end
  end
end

File: internal/paths/paths_test.go
package paths

import (
	"os"
	"path/filepath"
	"slices"
	"testing"
)

func TestDataHomeUsesEnvWhenSet(t *testing.T) {
	t.Setenv("VBRAIN_HOME", "/tmp/custom-vbrain")
	if got := DataHome(); got != "/tmp/custom-vbrain" {
		t.Fatalf("DataHome() = %q, want /tmp/custom-vbrain", got)
	}
}

func TestDataHomeDefaultsToHomeVbrainWhenEnvBlankAndNotInBase(t *testing.T) {
	home, _ := os.UserHomeDir()
	want := filepath.Join(home, "vbrain")

	// cwd in a dir without wiki/ → not a base → falls back to ~/vbrain.
	t.Chdir(t.TempDir())
	t.Setenv("VBRAIN_HOME", "")
	if got := DataHome(); got != want {
		t.Fatalf("blank env, fora de base: DataHome() = %q, want %q", got, want)
	}
}

func TestDataHomeUsesLocalBaseWhenEnvBlankAndRunningInBase(t *testing.T) {
	// cwd is a base (has wiki/) → use it, even without VBRAIN_HOME (fixes the
	// cloud where the checkout is the base).
	dir := t.TempDir()
	if err := os.MkdirAll(filepath.Join(dir, "wiki"), 0o755); err != nil {
		t.Fatal(err)
	}
	t.Chdir(dir)
	t.Setenv("VBRAIN_HOME", "")
	if got := DataHome(); got != dir {
		t.Fatalf("dentro de base: DataHome() = %q, want %q", got, dir)
	}
}

func TestSourceCheckoutIsNotABase(t *testing.T) {
	// A dir with wiki/ but also a go.mod is the vbrain source checkout, not a
	// base — otherwise `vbrain install` would bootstrap base assets into the
	// source tree and push them to the code remote.
	dir := t.TempDir()
	if err := os.MkdirAll(filepath.Join(dir, "wiki"), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(dir, "go.mod"), []byte("module x\n"), 0o644); err != nil {
		t.Fatal(err)
	}
	if IsBase(dir) {
		t.Fatal("source checkout (wiki/ + go.mod) must not be treated as a base")
	}
	t.Chdir(dir)
	t.Setenv("VBRAIN_HOME", "")
	home, _ := os.UserHomeDir()
	if got, want := DataHome(), filepath.Join(home, "vbrain"); got != want {
		t.Fatalf("inside source checkout: DataHome() = %q, want %q", got, want)
	}
}

func TestDerivedPathsAreUnderDataHome(t *testing.T) {
	dir := t.TempDir()
	t.Setenv("VBRAIN_HOME", dir)

	if got, want := RawDir(), filepath.Join(dir, "raw"); got != want {
		t.Errorf("RawDir() = %q, want %q", got, want)
	}
	if got, want := WikiDir(), filepath.Join(dir, "wiki"); got != want {
		t.Errorf("WikiDir() = %q, want %q", got, want)
	}
	if got, want := DBPath(), filepath.Join(dir, "db", "vbrain.sqlite3"); got != want {
		t.Errorf("DBPath() = %q, want %q", got, want)
	}
	if got, want := TmpDir(), filepath.Join(dir, "raw", ".tmp"); got != want {
		t.Errorf("TmpDir() = %q, want %q", got, want)
	}
}

func TestEnsureDirsCreatesFlatStructure(t *testing.T) {
	dir := t.TempDir()
	t.Setenv("VBRAIN_HOME", dir)

	if err := EnsureDirs(); err != nil {
		t.Fatal(err)
	}
	for _, sub := range []string{"raw", "wiki", "db", filepath.Join("raw", ".tmp"), filepath.Join("wiki", RealtimeDir)} {
		if fi, err := os.Stat(filepath.Join(dir, sub)); err != nil || !fi.IsDir() {
			t.Errorf("expected directory %q created", sub)
		}
	}
}

func TestEnsureDirsDoesNotCreateTypeFolders(t *testing.T) {
	dir := t.TempDir()
	t.Setenv("VBRAIN_HOME", dir)

	if err := EnsureDirs(); err != nil {
		t.Fatal(err)
	}
	for _, old := range []string{"concepts", "decisions", "gotchas", "notes", "_rules"} {
		if _, err := os.Stat(filepath.Join(dir, "wiki", old)); !os.IsNotExist(err) {
			t.Errorf("type folder %q should no longer be created (flat layout)", old)
		}
	}
}

func TestKindsIncludeAllSupportedMetadata(t *testing.T) {
	for _, k := range []string{"concept", "decision", "gotcha", "note", "rule", "realtime"} {
		if !slices.Contains(Kinds, k) {
			t.Errorf("Kinds should contain %q", k)
		}
	}
}
